### DISCUSSION TITLE
Allow F rank as valid score

### DISCUSSION THREAD

**Discussion by alexpado** [ppy/osu#24382] [Ideas]
### Type

Crash to desktop

### Bug description

I made a mod that is like NF, but with a score multiplier of x1.0, allowing you to beat your personal best and keep it in your local score with a nice "F" if you reached 0 HP at any point in the map. This can allow the player to see which score they can aim for without failing.

The score does save well, the replay is available without issue, but at the end of the play (replay and gameplay), the game crashes because of this:

https://github.com/ppy/osu/blob/057bac39dd0797a42fd39c8ccbd24353d00189bf/osu.Game/Screens/Play/Player.cs#L785

While my mod can be questionnable, I think having completed play with a F rank should be allowed as any failed replay is allowed anyway, and are using the F rank for this specific purpose.

I commented this line locally and the result screen is displayed without any visible issue.

Is there any specific reason for this line to be here that I'm missing ?

### Screenshots or videos

# The error
![image](https://github.com/ppy/osu/assets/31733699/70041dc6-e6aa-44ec-9466-82848b18739f)

# In Game saved score
![image](https://github.com/ppy/osu/assets/31733699/225aaf27-2bce-458a-ab26-65bf28eef778)

# While having the assertion commented out
![image](https://github.com/ppy/osu/assets/31733699/9ce218d7-c349-4519-bb5b-97b65026d55e)

### Version

057bac3 (master)

### Logs

(see first screenshot reference)

**Comment by alexpado** [comment]
As a sidenote, this could also be used for taiko. When the HP target isn't reached, instead of showing the fail screen, showing the score with the F rank could be possible

**Comment by bdach** [comment]
This is a feature proposal and not a bug, and as such I've converted this into a discussion. It's certainly not a crash to desktop.

More materially:

> I made a mod that is like NF, but with a score multiplier of x1.0, allowing you to beat your personal best and keep it in your local score with a nice "F" if you reached 0 HP at any point in the map. This can allow the player to see which score they can aim for without failing.

This has the same problem as the numerous existing discussions suggesting to make NF have the multiplier of x1.0, in that there is no reason to not have the mod always enabled. Maybe this could be salvaged if the score didn't submit online, but lazer no longer has the notion of a "ranked mod" anymore, so this isn't really doable. This mod would be the sole exception again.

**Reply by alexpado** [reply]
I agree for it to be considered a discussion, but it seems more like a crash to me. Without debugging, the game abruptly closes without any warning or error message, returning to the desktop.

However, I could be misinterpreting "crash to desktop" though.

Edit for your edit:

Well, again, this could be used for Taiko to keep failed play too, without even talking about mods

**Reply by bdach** [reply]
> I agree for it to be considered a discussion, but it seems more like a crash to me. Without debugging, the game abruptly closes without any warning or error message, returning to the desktop.

It's a "crash" in a mod that you wrote that hasn't been PRed even, let alone merged into master. No repository will consider this a valid bug.

The assertion in question is valid on `master`. The fact that your mod breaks that assertion (because it's fundamentally changing core assumptions) does not change that.

**Reply by alexpado** [reply]
Thanks for the clarification.

Well, if the only reason this assertion exists is to not submit failed score, could the assertion be moved to the submit part of osu! ?

I feel like having the ability to save replays with F rank but not complete score with F rank is weird, even if in the core concept I understand why it has been placed there in the first place

**Reply by bdach** [reply]
The assertion is there to ensure that no gameplay session (either live, or replay) progresses from a failed score to the results screen, as the commit message of https://github.com/ppy/osu/commit/4dd0c2c7a5eb3b0d57fbed9bdf4acf1bacb439e1 states clearly. Results screen can be shown for F rank replays from song select, but not from gameplay/replay playback.

Whether this is to change is one of the core points in this entire discussion. If the proposal is rejected then there is no reason to change the assertion.

**Reply by alexpado** [reply]
The commit description state:
> but I don't want this to cause crashes for an end user if it does happen to occur.

It seems that the assertion is here as a temporary mesure and something else is planned for the future. Let's wait for peppy I guess

**Reply by bdach** [reply]
It's not a temporary measure. It's supposed to let us catch these cases during development, but not fail in release configurations, since an F rank showing on results screen in some scenario we haven't anticipated is preferable to the game crashing.

**Reply by alexpado** [reply]
F rank does already show naturally in the song select, as you can save failed replay. That's why I don't understand why complete play aren't allowed, as they are basically "failed" replay the goes at the end of the song

Also, just for information: I've been playing with this line commented out and having lot of F and did not experienced any crashes

**Comment by alexpado** [comment]
After further observation of the source code for the score display:

1. The color for the `F` rank is the same a the `D` rank, as both fall in the `default` branch in `DrawableRank.getRankNameColour()`
2. The `ScoreRank.GetDescription()` description fallback (I guess) to the `toString()` method, just like any other ranks (for `GetRankName` method still in `DrawableRank.cs`
3. Same as 1 regarding `OsuColour.ForRank(rank)`, it goes into `default` branch.

So, regarding the score display, I don't see any issue as it seems to be handled just like the `D` rank (even with the little animation to rotate it)

I'm not trying to force it upon the team, but I'm genuinely curious why the `F` rank could be so bad to allow.

The only "real" issue I could see there is the TODO left
https://github.com/ppy/osu/blob/22163020c348c8ad1545b95d33665277425f4a1f/osu.Game/Scoring/ScoreRank.cs#L12-L14

Which in itself shouldn't be a big hassle to do

**Reply by bdach** [reply]
It's not about whether "F rank should be allowed or not". It already is. You can locally import failed plays.

The actual questions that your proposed mod is raising are:
- whether it should be possible to reach the results screen after a failed play
- whether F rank plays achieved with your mod should be submittable to online leaderboards
- whether failed plays with this mod should have a 1.0x score multiplier as proposed

**Reply by bdach** [reply]
> The color for the `F` rank is the same a the `D` rank, as both fall in the `default` branch in `DrawableRank.getRankNameColour()`

And for reference this is probably wrong, it's only because the F rank was added bypassing designer input.

**Reply by alexpado** [reply]
> whether it should be possible to reach the results screen after a failed play

~~Imo, it should be possible to, to see the detail about your play, as player might be interested how much Great or Meh they did without having to add skin element and pausing the replay right before the failure screen to see those stats.~~

I forgot to turn my brain on when I wrote this, saving the replay allow you to access such stats from the leaderboard, but really, what's the difference between accessing score from the leaderboard and accessing it from a finished play ?

> whether F rank plays achieved with your mod should be submittable to online leaderboards

I think everybody could agree that it should not, as it would count toward all account stats, which is not favorable in many way (and would add even more data server side !).

The idea would be to consider F ranked play as more like a failed replay rather than "a map that has been finished and has a score" (we would be in the middleground)

> whether failed plays should have a 1.0x score multiplier as proposed

The issue is not about the score multiplier, nor the NF mod in itself, but I think that even for other ruleset (custom or not), having F result screen could be great.

**I'm not asking to modify the NF here, it's all about the F rank result screen made possible.** It could even never appear in the base game but made possible for external ruleset only.
